### PR TITLE
Add new rule: no-obsolete-elements

### PR DESCRIPTION
**Pull request by MelSumner**
If merged, this PR adds a new rule to check for obsolete elements and throw an error if they are used.

### CODE REVIEW

File: docs/rule/no-obsolete-elements.md [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+This rule forbids the use of obsolete elements.
+
+### Examples
+
+This rule **forbids** the following:
+
+```hbs
+  <applet></applet>
+  <acronym></acronym>
```
Comment (rwjblue) [inline docs/rule/no-obsolete-elements.md]: There is some odd indentation in this list, can you make them all indented the same way?

File: docs/rule/no-obsolete-elements.md [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+  <marquee></marquee>
+  <multicol></multicol>
+  <nobr></nobr>
+  <spacer></spacer>
+  <tt></tt>
+```
+
+This rule **allows** the following:
+
+```hbs
```
Comment (rwjblue) [inline docs/rule/no-obsolete-elements.md]: I'd probably remove this listing, its nearly impossible to be exhaustive and the main point is "it works with everything else"

File: lib/rules/lint-no-obsolete-elements.js [hunk 3]
```diff
@@ -0,0 +1,53 @@
+'use strict';
+
+const Rule = require('./base');
+
+const errorMessage = element => `Use of <${element}> detected. Do not use deprecated elements.`;
+
+const OBSOLETE_ELEMENTS = [
```
Comment (rwjblue) [inline lib/rules/lint-no-obsolete-elements.js]: Can you put an inline comment just above here with the link to the whatwg listing?

File: lib/rules/lint-no-obsolete-elements.js [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+  'nobr',
+  'spacer',
+  'tt',
+];
+
+module.exports = class NoObsoleteElements extends Rule {
+  visitor() {
+    return {
+      ElementNode(node) {
+        if (OBSOLETE_ELEMENTS.includes(node.tag)) {
```
Comment (rwjblue) [inline lib/rules/lint-no-obsolete-elements.js]: We also need to check if the `node.tag` is a block param.

Can you add a "good" test for:

```hbs
{{#let (component 'whatever-here') as |plaintext|}}
  <plaintext />
{{/let}}
```

^ breaks the `no-shadowed-elements` rule, but should **not** fail this rule

File: yarn.lock [hunk 5]
```diff
@@ -96,42 +96,34 @@
   integrity sha512-tHq6qdbT9U1IRSGf14CL0pUlULksvY9OZ+5eEgl1N7t+OA3tGvNpxJCzuKQlsNgCVwbAs670L1vcVQi8j9HjnA==
 
 "@types/node@*":
-  version "10.12.18"
-  resolved "https://registry.yarnpkg.com/@types/node/-/node-10.12.18.tgz#1d3ca764718915584fcd9f6344621b7672665c67"
-  integrity sha512-fh+pAqt4xRzPfqA6eh3Z2y6fyZavRIumvjhaCL753+TVkGKGhpPeyrJG2JftD0T9q4GF00KjefsQ+PQNDdWQaQ==
+  version "12.0.4"
```
Comment (rwjblue) [inline yarn.lock]: Can you reset the changes to this file? It seems unrelated to adding the rule.

If `yarn.lock` on master is out of sync we should update in a separate (focused) PR.

File: lib/rules/lint-no-obsolete-elements.js (hunk 4 above)
Comment (MelSumner) [reply lib/rules/lint-no-obsolete-elements.js]: @rwjblue just to be clear- we both want to prevent let-as-(obsolete element), right?

File: lib/rules/lint-no-obsolete-elements.js [hunk 6]
```diff
@@ -0,0 +1,54 @@
+'use strict';
+
+const Rule = require('./base');
+
+const errorMessage = element => `Use of <${element}> detected. Do not use deprecated elements.`;
```
Comment (bmish) [inline lib/rules/lint-no-obsolete-elements.js]: I would recommend exporting this error message function and then importing it from the test file to avoid duplication. See `no-negated-condition` rule files for an example.

File: test/unit/rules/lint-no-obsolete-elements-test.js [hunk 7]
```diff
@@ -0,0 +1,258 @@
...
+const generateRuleTests = require('../../helpers/rule-test-harness');
+
+generateRuleTests({
+  name: 'no-obsolete-elements',
+
+  config: true,
+
+  good: ['<div></div>'],
+
+  bad: [
```
Comment (bmish) [inline test/unit/rules/lint-no-obsolete-elements-test.js]: This long array of test cases is pretty tedious, and of course duplicates the list of deprecated elements from the rule file, making it more work to update the list later. Instead, you could export the `OBSOLETE_ELEMENTS` array from the rule file, and then import it here to generate the test cases via a `forEach` loop.

File: lib/rules/lint-no-obsolete-elements.js (hunk 4 above)
Comment (rwjblue) [reply lib/rules/lint-no-obsolete-elements.js]: Kinda, yes. _This_ rule should **not** flag any block params at all, but the no-shadowed-elements rule prevents using _any_ lowercased block param names in the element position. 

So under the no-shadowed-element rule (which is enabled by default), the example I gave above would be disallowed.

File: lib/rules/lint-no-obsolete-elements.js (hunk 6 above)
Comment (rwjblue) [reply lib/rules/lint-no-obsolete-elements.js]: Ya, good idea

File: test/unit/rules/lint-no-obsolete-elements-test.js (hunk 7 above)
Comment (rwjblue) [reply test/unit/rules/lint-no-obsolete-elements-test.js]: Ya, something like:

```js
bad: OBSOLETE_ELEMENTS.map(element => {
  return {
    /* generated POJO for this element */
  }
}),
```

File: lib/rules/lint-no-obsolete-elements.js (hunk 4 above)
Comment (MelSumner) [reply lib/rules/lint-no-obsolete-elements.js]: Hmmm. I think it should block use of these elements, block params or not. Let's schedule a call to discuss more, so I can understand the logic.

Comment (rwjblue) [reply lib/rules/lint-no-obsolete-elements.js]: Seems good, but we won't be landing this rule until we have tests ensuring that this example does not fail the new rule being added:

```hbs
{{#let (component 'whatever-here') as |plaintext|}}
  <plaintext />
{{/let}}
```

Comment (MelSumner) [reply lib/rules/lint-no-obsolete-elements.js]: I see. I still would like to understand it better. Thanks! :+1:

File: docs/rule/no-obsolete-elements.md [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+<center></center>
+<font></font>
+<marquee></marquee>
+<multicol></multicol>
+<nobr></nobr>
+<spacer></spacer>
+<tt></tt>
+```
+
+This rule **allows** the following:
```
Comment (rwjblue) [inline docs/rule/no-obsolete-elements.md]: Can probably remove this line completely, and bold allows in the sentence just below.

**Comment by rwjblue** [comment]
Thank you!
